data/code/lib.c

```diff
@@ -216,3 +217,185 @@ int itemshufflemgr_calcslot_ex(const unsigned int* const itemshufflemgr,
 }
 
 #endif
+
+#if SECTIONED_COURSES
+
+static char passed;
+static unsigned int sectionCount = 0;
+
+// override the lap count in a section course to be the number of section points
+void section_lap_override()
+{
+    register unsigned short reg3 asm ("r3");
+    register unsigned short reg9 asm ("r9");
+    asm("or %r0,%r3,%r3 \t\n");
+
+    reg9 = (unsigned short)sectionCount;
+
+    if (reg9 != 0)
+    {
+        if (reg9 > 9)
+        {
+            asm("li %r10, 0x9        \t\n"
+                "sth %r10,0x2e(%r31) \t\n");
+        }
+        else
+        {
+            asm("sth %r9, 0x2e(%r31)");
+        }
+    }
+}
+
+void reset_sect_count()
+{
+    asm("or %r31, %r3, %r3");  // run hijacked instruction
+    sectionCount = 0;
+}
+
+// count the section points to jury-rig the "max laps" count to always be the section number
+void count_sect_points()
+{
+    register int *reg0 asm ("r0");
+    register int reg3 asm ("r3");
+
+    asm("stw %r4, 0x8(%r31)                 \t\n"); // run hijacked instruction
+
+    // is the "section point" bit set?
+    asm("lbz   %r9, 0x0018 (%r4)            \t\n" // r9 = *(r4 + 0x18) ("shortcut point" flag)
+        "subic %r10, %r9, 0x01	            \t\n"
+        "subfe %r9, %r10, %r9	            \t\n"
+        "rlwinm.  %r10, %r9, 0x0, 0x18, 0x1f\t\n" // cast to byte
+        "bnelr                              \t\n" // skip invalid section points
+        "lbz   %r9, 0x001B (%r4)            \t\n" // r9 = *(r4 + 0x1B) ("section checkpoint" flag)
+        "subic %r10, %r9, 0x01	            \t\n"
+        "subfe %r9, %r10, %r9	            \t\n"
+        "rlwinm.  %r10, %r9, 0x0, 0x18, 0x1f\t\n" // cast to byte
+        "beqlr                              \t\n" // skip below if sections aren't set
+        );
+
+    ++sectionCount;
+}
+
+void set_force_lap()
+{
+    register char reg0 asm ("r0");
+    register char reg9 asm ("r9");
+
+    bool isSection;
+
+    passed = (char)0;
```

Comment:
Why is `passed` a static variable?

We can define some _local_ variables if it helps readability, but they should be marked as constant, and defined in one go:

```cpp
const bool passed = (bool)reg9;
const bool is_section = reg0 != '\0';

if (passed && is_section)
{
    force_lap_increment();
}
```